engine/kv_engine_cleaner.cpp

```diff
@@ -235,366 +237,431 @@ void KVEngine::cleanNoHashIndexedSkiplist(
+  }
+  if (outdated_dl_records.size() > 0) {
+    outdated_records_num += outdated_dl_records.size();
+    std::vector<DLRecord*> to_free_dls;
+    version_controller_.UpdateLocalOldestSnapshot();
+    auto release_time = version_controller_.LocalOldestSnapshotTS();
+    auto iter = outdated_dl_records.begin();
+    while (iter != outdated_dl_records.end()) {
+      if (iter->release_time < release_time) {
+        to_free_dls.emplace_back(iter->record);
+      } else {
+        purge_dl_records.push_back(iter->record);
+      }
+      iter++;
+    }
+    pending_clean_records.pending_purge_dls.emplace_front(
+        std::move(to_free_dls), release_time);
+  }
+  return outdated_records_num;
+}
 
-    std::vector<StringRecord*> purge_string_records;
-    std::vector<DLRecord*> purge_dl_records;
-
-    // Iterate hash table
-    auto hashtable_iter =
-        hash_table_->GetIterator(start_slot_idx, end_slot_idx);
-    while (hashtable_iter.Valid()) {
-      if (slot_num++ % kSlotSegment == 0) {
-        // Clean cached
-        {
-          size_t i = round_robin_id_.fetch_add(1) % configs_.max_access_threads;
-          auto& tc = cleaner_thread_cache_[i];
-          std::deque<CleanerThreadCache::OutdatedRecord<StringRecord>>
-              outdated_string_records;
-          std::deque<CleanerThreadCache::OutdatedRecord<DLRecord>>
-              outdated_dl_records;
-          if (tc.outdated_dl_records.size() > kMaxCachedOldRecords ||
-              tc.outdated_string_records.size() > kMaxCachedOldRecords) {
-            std::lock_guard<SpinMutex> lg(tc.mtx);
-            if (tc.outdated_string_records.size() > kMaxCachedOldRecords) {
-              outdated_string_records.swap(tc.outdated_string_records);
-            }
-            if (tc.outdated_dl_records.size() > kMaxCachedOldRecords) {
-              outdated_dl_records.swap(tc.outdated_dl_records);
-            }
-          }
-          if (outdated_string_records.size() > 0) {
-            std::vector<StringRecord*> to_free_strings;
-            version_controller_.UpdateLocalOldestSnapshot();
-            auto release_time = version_controller_.LocalOldestSnapshotTS();
-            auto iter = outdated_string_records.begin();
-            while (iter != outdated_string_records.end()) {
-              if (iter->release_time < release_time) {
-                to_free_strings.emplace_back(iter->record);
-              } else {
-                purge_string_records.push_back(iter->record);
+static inline int64_t PushToOutdatedPool(
+    StringRecord* old_record, std::vector<StringRecord*> purge_string_records) {
+  if (old_record) {
+    purge_string_records.emplace_back(old_record);
+    return 1;
+  }
+  return 0;
+}
+
+static inline int64_t PushToOutdatedPool(
+    DLRecord* old_record, std::vector<DLRecord*> purge_dl_records) {
+  if (old_record) {
+    purge_dl_records.emplace_back(old_record);
+    return 1;
+  }
+  return 0;
+}
+
+double KVEngine::cleanSlotBlockOutDated(
```

Comment:
I think slot block is a little confused, maybe just cleanHashtableOutdated(start_slot, num_slot) ?

Reply:
keep before: cleanoutdated?

Reply:
ok